### ISSUE TITLE
Question about preview images

### ISSUE THREAD

**Issue by vinerz** [shaoanlu/faceswap-GAN#28]
First os all, thank you so much for such a detailed notebook!

Could you explain a little bit more about the training previews?

I am a long time developer, but my specialty is at node.js.
With this new face-swapping hype, I decided to jump on the train for fun, but some stuff is still a bit confusing for me.
I've read some GAN papers and I found this to be the most effective (and most fun!) path to do face swapping.

**First of all:** Why are the faces on the output sample blue-ish? Is this the correct behavior?
**Second:** I'm noticing some weird hard pixels around the actors' noses, specifically around Tom Hiddleston's nose. Is this also supposed to happen?
**Third:** The third column at the sample masks is empty. Is this correct?

**Hardware**
I am currently running the training script on this setup:

- **NVIDIA GK210 GPU w/ 12GiB DDR5 VRAM**
- tensorflow 1.5.0
- CUDA 9.0

This is running ~1 iteration/sec.
Current loss information:

`Loss_DA: 0.001124 Loss_DB: 0.000390 Loss_GA: 0.008428 Loss_GB: 0.010443`

**Sample after ~1k iterations**

![_sample_faces](https://user-images.githubusercontent.com/1482034/35488392-530ba01c-046f-11e8-8b3e-ec5c43e2c604.jpg)

**Sample after ~2.5k iterations**

![_sample_faces-3](https://user-images.githubusercontent.com/1482034/35488399-7aba7624-046f-11e8-8ef8-2d38e2e93e09.jpg)

**Most recent sample after ~3.2k iterations with mask preview**

![_sample_faces-5](https://user-images.githubusercontent.com/1482034/35488447-48f436f6-0470-11e8-924e-ec727ef5ddd8.jpg)
![_sample_masks](https://user-images.githubusercontent.com/1482034/35488483-f9ea510c-0470-11e8-8e93-24e65038afcb.jpg)

**Comment by vinerz** [comment]
**Update:** For some reason, the third column decided to show up, from nothing.

![_sample_masks-2](https://user-images.githubusercontent.com/1482034/35488562-370b847e-0472-11e8-9307-d672d956d26c.jpg)

**Comment by shaoanlu** [comment]
1. The output of generator has channel ordering BGR instead of RGB. Probably there is a missing line `cv2.cvtColor(...)` before displaying function. Which jupyter notebook did you use?

For 2 and 3, they have similar cause that the generator (auto-encoder) is not trained enough. Outputs of generator are too blurry that there is little backprop. information from discriminator being able to flow through generator. In other words, the predicted mask will converge only after autoencoder can produce good result.

**Comment by vinerz** [comment]
Hey @shaoanlu! Thanks for the explanation 😄

Found the _Smurfs_ issue:
I changed `plt.imshow` to `cv2.imwrite` in order to save previews as JPEGs and it uses the infamous BGR channels.
Removed the channel conversion and it works like a charm now!

**Here's a preview:**
![_sample_faces-7](https://user-images.githubusercontent.com/1482034/35519244-d9669474-04fa-11e8-87ce-bb35c3c032be.jpg)
![_sample_masks-3](https://user-images.githubusercontent.com/1482034/35519258-e33468aa-04fa-11e8-87d6-426a9ee324fa.jpg)

About the training time:

I've also trained the same images with the regular deepfake algorithm and after 14 hours of traning, I got some pretty decent results, as shown below:

![sample_regular-5](https://user-images.githubusercontent.com/1482034/35519395-3aee8166-04fb-11e8-840d-7818b9a403ac.jpg)

What I've read about GAN training is that it takes a while to train, but once the network reaches a certain point, the generator converges and increases its accuracy exponentially. Is this correct?

**Comment by shaoanlu** [comment]
Yes, it is also common that there are weird artifacts on intermediate preview faces. To leverage the power of discriminator, the generator has to first learn well on how to reconstruct warped faces back to unwarped ones (reconstruction L1 loss should be small enough). Then the discriminator can send instructive signal to generator for weights update.
